File: QytetetJava/src/modeloqytetet/Especulador.java

package modeloqytetet;


public class Especulador extends Jugador {
    private int fianza;
    
    
    // Constructor de especulador
    protected Especulador (Jugador jugador, int fianza){
        super (jugador);
        this.fianza = fianza;
    }
    
    // Método para pagar impuesto 
    @Override
    protected void pagarImpuesto (){
        modificarSaldo (-(getCasillaActual().getCoste()/2));
    }
    
    // Método para consultar si el especulador debe ir a la cárcel
    
    @Override
    protected boolean deboIrACarcel(){
        boolean debe_ir = false;
        debe_ir = super.deboIrACarcel();
        if (!debe_ir)
            debe_ir = !pagarFianza();
        return debe_ir;
    }
    
    // Método privado para pagar la fianza
    
    private boolean pagarFianza(){
        boolean pagada = false;
        
        if (getSaldo() > fianza){
            modificarSaldo(-fianza);
            pagada = true;
        }
        return pagada;
    }
    
    // Método para obtener el factor de especulador
//    @Override
//    protected int getFactorEspeculador(){}
    
    // Método para convertirse
    @Override
    protected Especulador convertirme (int fianza){
        return this;
    }

    // Método que sobreescribe el de jugador y devuelve si puede edificar una casa
    
    @Override
    protected boolean puedoEdificarCasa(TituloPropiedad titulo){
        return tengoSaldo(titulo.getPrecioEdificar()) && titulo.getNumCasas() < 8 && titulo.getNumHoteles() == 0;
    }
    
    // Método que sobreescribe el de jugador y devuelve si puede edificar una casa
    
    @Override
    protected boolean puedoEdificarHotel(TituloPropiedad titulo){
        boolean tengo_suficientes_casas = titulo.getNumCasas() == 8;
        return tengoSaldo(titulo.getPrecioEdificar()) && ((titulo.getNumHoteles() == 0 && tengo_suficientes_casas)
                || (titulo.getNumHoteles() < 8 && titulo.getNumHoteles() > 0));
    }
    
    @Override
    public String toString(){
        String aDevolver = super.toString().substring(0, super.toString().length()-1);
        aDevolver += ", Fianza= ";
        aDevolver += fianza;
        
        return aDevolver + '}';
    }
    
}

File: QytetetRuby/lib/ModeloQytetet/especulador.rb
# encoding: utf-8
require_relative "../ModeloQytetet/qytetet"
require_relative "../ControladorQytetet/opcion_menu"
require_relative "../ControladorQytetet/controlador_qytetet"
require_relative "../ModeloQytetet/calle"
require_relative "../ModeloQytetet/casilla"
require_relative "../ModeloQytetet/dado"
require_relative "../ModeloQytetet/especulador"
require_relative "../ModeloQytetet/estado_juego"
require_relative "../ModeloQytetet/jugador"
require_relative "../ModeloQytetet/metodo_salir_carcel"
require_relative "../ModeloQytetet/otra_casilla"
require_relative "../ModeloQytetet/sorpresa"
require_relative "../ModeloQytetet/sorpresa"
require_relative "../ModeloQytetet/tablero"
require_relative "../ModeloQytetet/tipo_casilla"
require_relative "../ModeloQytetet/tipo_sorpresa"
require_relative "../ModeloQytetet/titulo_propiedad"
module ModeloQytetet
  class Especulador < Jugador
    attr_accessor :fianza
    def self.copia (jug, fianza)
      j = super(jug)
      j.fianza = fianza
      return j
    end
 
    def convertirme (fianza)
      return self
    end
  
    # Un especulador paga la mitad de impuestos que un jugador normal
    def pagar_impuesto
      @saldo -= (@casilla_actual.coste / 2)
    end 
  
    # Devuelve false si no tiene saldo suficiente para pagar la fianza
    # devuelve true si la paga
  
    def pagar_fianza
      if @saldo >= @fianza
        @saldo -= @fianza
        return true
      end
  
      return false
    end
  
    # Devuelve true si no tiene carta libertad y no puede pagar la fianza
  
    def debo_ir_a_carcel
      if (super)
        return !pagar_fianza
      else
        return true 
        # Lo hemos hecho así, porque pensamos que si tiene carta y llama a 
        # pagar fianza y tiene el saldo suficiente, se restaría dinero
        # del saldo aún teniendo la carta libertad
      end
    end
  
    # Devuelve true si puede edificar una casa en una propiedad
    def puedo_edificar_casa (titulo)
      return @saldo >= titulo.precio_edificar && titulo.num_casas < 8 && titulo.num_hoteles == 0 
    end
   
    # Devuelve true si puede edificar un hotel en una propiedad
    def puedo_edificar_hotel (titulo)
      tengo_suficientes_casas = titulo.num_casas == 8
      return (@saldo >= titulo.precio_edificar) && ((titulo.num_hoteles == 0 && tengo_suficientes_casas) || (titulo.num_hoteles < 8 && titulo.num_hoteles > 0))
    end
 
    private :pagar_fianza
   
    # Añade al to_s de la superclase el campo fianza
  
    def to_s
      to_return = super
    
      to_return += "\nFianza: #{@fianza}"
    
      return to_return
    end
  end
end